library/core/src/ptr/mod.rs

```diff
@@ -65,6 +65,12 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//! It is guaranteed that an allocated object never spans more than `isize::MAX` bytes.
+//! The address "at the end" of an allocated object (i.e., one past the last byte that
+//! is contained inside the allocated object) can always be computed without wrapping
+//! around the address space. Equivalently, the address `usize::MAX` is never contained
+//! in an allocated object.
```

Comment:
I prefer this phrasing which expresses the requirements more clearly and covers the null case as well:

> An allocated object cannot contain the first (`0`) and last (`usize::MAX`) bytes of the address space.

Reply:
I think it should definitely say that the "at the end" pointer can be computed without wrapping, but I'm fine with stating this as a consequence of the allocated object not containing `usize::MAX`.

Reply:
Updated. It currently doesn't mention the first address; see https://github.com/rust-lang/rust/pull/116675#discussion_r1364188415.